telemetry/run.py

```diff
@@ -0,0 +1,97 @@
+    Returns a dict, whose keys are (group) name of the sensors and values are pairs of (label x data).
+    '''
+    current_group = 'ungrouped'
+    result = dict({ current_group : [] })
+    for line in info:
+        # omit empty lines and `N/A` data
+        if line and not 'N/A' in line:
+            # Adapter names does not contain spaces
+            if ' ' not in line:
+                current_group = line
+                result.update({ line : [] })
+            else:
+                label, data = ' '.join(line.split()).split()[:2]
+                result[current_group].append((label.replace(':', ''), ''.join(data)))
+    # return with empty entries removed
+    return dict(filter(lambda x: len(x[1]), result.items()))
+
+def parse_gpu_stat(info:list) -> list:
+    '''
+    Takes data returned by `nvidia-smi`, which contains a string with two lines, returns
+    the last line, which is the data we need to monitor during an experiment.
+    '''
+    filtered = list(filter(lambda x: x, info))
+    # ensure the list is not empty
+    assert filtered
+    return filtered[-1].split(', ')
+
+def start_job(fp_dir, nvidia_fields, time_span, time_run) -> None:
+    '''
+    A chornological job, runs every `time_span` seconds.
+    Fetches data from `nvidia-smi` and `sensors`, then parse it
+    and write to corresponding files. Data will not be processed
+    during monitoring in order to minimize the influence to the running experiments.
+
+    Note: The process will be halted by `dashboard.py` when an experiment ends. 
+    '''
+    threading.Timer(time_span, start_job, args=[fp_dir, nvidia_fields, time_span, time_run + 1]).start()
+    nvidia_smi = subprocess.run(['nvidia-smi', '--format=csv', '--query-gpu={}'.format(','.join(nvidia_fields))],
+                            stdout=subprocess.PIPE, timeout=10)
+    parsed_data = parse_gpu_stat(nvidia_smi.stdout.decode().split('\n'))
+    # The length of lists of data should be the same
+    # since we are using `nvidia_fields` to call the command
+    assert len(parsed_data) == len(nvidia_fields)
+    # timestamp = parsed_data[0]
+    time_after = time_span * time_run
+    for filename, data in zip(nvidia_fields[1:], parsed_data[1:]):
+        with open(os.path.join(fp_dir, 'gpu', filename), 'a+') as fp:
+            # fp.write(f'{timestamp[:-4]} {data}\n')
+            fp.write(f'{time_after} {data}\n')
+
+    sensors = subprocess.run(['sensors'], stdout=subprocess.PIPE)
+    cpu_stat = parse_cpu_stat(sensors.stdout.decode().split('\n'))
+    # timestamp = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')
+    for (fname, entries) in cpu_stat.items():
+        if entries[1:]:
+            with open(os.path.join(fp_dir, 'cpu', fname), 'a+') as fp:
+                for (label, data) in entries[1:]:
+                    # fp.write(f'{timestamp} {label} {data}\n')
+                    fp.write(f'{time_after} {label} {data}\n')
+
+def main(args):
```

Comment:
Something you could do to make your life easier would be to make use of some of the functions in `shared`, such as `invoke_main` and the `DashboardInfo` class. This way, you could reduce the boilerplate and centralize the logic that has to do with figuring out directories (very error-prone)

Reply:
I'm trying to do so. One problem is that to enable modules in `shared`, I have to run the process with a shell script. This make it difficult to terminate the process whlie an experiment ends since I cannot get the `pid` or the process instant of the actual python process if I run it with a shell script.

Reply:
You can't just terminate the outer script?

Another strategy to implementing the telemetry process would be to include everything as a library in `shared` or somewhere in the Pythonpath and use a function from `threading` to dispatch it _from_ the dashboard.

Reply:
Terminating the outer script can't stop the python process invoked in the script. I'll try to integrate it as a library. Thanks for the suggestion!

Reply:
You might actually be right to do it as a separate process, though. Apparently there is some weirdness in Python about threading that I didn't know about, so maybe a little code duplication is for the best

Reply:
You might be able to set the Pythonpath in the Python command you launch though

Reply:
@AD1024 yes, you can set the `env` argument in `subprocess.run()` or `subprocess.Popen()` so you can add the shared directory to the Pythonpath (you can use `os.environ` for that)

Reply:
Thanks for the information, I'll try that.